File: php/features.php
<?php
declare(strict_types=1);

// Univec SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/AuditFeature.php';
require_once __DIR__ . '/feature/CacheFeature.php';
require_once __DIR__ . '/feature/ClienttrackFeature.php';
require_once __DIR__ . '/feature/CostFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/LogFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/NetsimFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/ProxyFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RbacFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/SecretsFeature.php';
require_once __DIR__ . '/feature/StreamingFeature.php';
require_once __DIR__ . '/feature/TelemetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class UnivecFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new UnivecBaseFeature();
            case "audit":
                return new UnivecAuditFeature();
            case "cache":
                return new UnivecCacheFeature();
            case "clienttrack":
                return new UnivecClienttrackFeature();
            case "cost":
                return new UnivecCostFeature();
            case "debug":
                return new UnivecDebugFeature();
            case "idempotency":
                return new UnivecIdempotencyFeature();
            case "log":
                return new UnivecLogFeature();
            case "metrics":
                return new UnivecMetricsFeature();
            case "netsim":
                return new UnivecNetsimFeature();
            case "paging":
                return new UnivecPagingFeature();
            case "proxy":
                return new UnivecProxyFeature();
            case "ratelimit":
                return new UnivecRatelimitFeature();
            case "rbac":
                return new UnivecRbacFeature();
            case "retry":
                return new UnivecRetryFeature();
            case "secrets":
                return new UnivecSecretsFeature();
            case "streaming":
                return new UnivecStreamingFeature();
            case "telemetry":
                return new UnivecTelemetryFeature();
            case "test":
                return new UnivecTestFeature();
            case "timeout":
                return new UnivecTimeoutFeature();
            default:
                return new UnivecBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "audit":
            case "cache":
            case "clienttrack":
            case "cost":
            case "debug":
            case "idempotency":
            case "log":
            case "metrics":
            case "netsim":
            case "paging":
            case "proxy":
            case "ratelimit":
            case "rbac":
            case "retry":
            case "secrets":
            case "streaming":
            case "telemetry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: rb/features.rb
# Univec SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/audit_feature'
require_relative 'feature/cache_feature'
require_relative 'feature/clienttrack_feature'
require_relative 'feature/cost_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/log_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/netsim_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/proxy_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/rbac_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/streaming_feature'
require_relative 'feature/telemetry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module UnivecFeatures
  def self.make_feature(name)
    case name
    when "base"
      UnivecBaseFeature.new
    when "audit"
      UnivecAuditFeature.new
    when "cache"
      UnivecCacheFeature.new
    when "clienttrack"
      UnivecClienttrackFeature.new
    when "cost"
      UnivecCostFeature.new
    when "debug"
      UnivecDebugFeature.new
    when "idempotency"
      UnivecIdempotencyFeature.new
    when "log"
      UnivecLogFeature.new
    when "metrics"
      UnivecMetricsFeature.new
    when "netsim"
      UnivecNetsimFeature.new
    when "paging"
      UnivecPagingFeature.new
    when "proxy"
      UnivecProxyFeature.new
    when "ratelimit"
      UnivecRatelimitFeature.new
    when "rbac"
      UnivecRbacFeature.new
    when "retry"
      UnivecRetryFeature.new
    when "streaming"
      UnivecStreamingFeature.new
    when "telemetry"
      UnivecTelemetryFeature.new
    when "test"
      UnivecTestFeature.new
    when "timeout"
      UnivecTimeoutFeature.new
    else
      UnivecBaseFeature.new
    end
  end
end
